packages/runtime/runtime-utils/src/dataStoreHelpers.ts

```diff
@@ -44,31 +33,33 @@ export function exceptionToResponse(err: any): IResponse {
             mimeType: "text/plain",
             status: responseErr.code,
             value: responseErr.message,
-            stack: responseErr.stack ?? getStack(),
+            get stack() { return responseErr.stack; },
         };
     }
+
+    // Capture error objects, not stack itself, as stack retrieval is very expensive operation, so we delay it
+    const errWithStack = generateErrorWithStack();
+
     return {
         mimeType: "text/plain",
         status,
         value: `${err}`,
-        stack: getStack(),
+        get stack() { return ((err?.stack) as (string|undefined)) ?? errWithStack.stack; },
     };
 }
 
-export function responseToException(response: IResponse, request: IRequest) {
+export function responseToException(response: IResponse, request: IRequest): Error {
```

Comment:
Is this function under test?  Code looks right but also the kind of diff where a bug could slip in...

Reply:
Oh I see your UT comment.  Good!